.eslintrc.json

```diff
@@ -234,7 +264,13 @@
                     "off"
                 ],
                 "no-await-in-loop": [
-                    "off"
+                    "error"
+                ],
+                "no-return-await": [
+                    "error"
+                ],
+                "no-promise-executor-return": [
+                    "error"
```

Comment:
This one I'm not that sure about. As I understand it'd prevent returning a promise from executor function, right? In that case I'd be against this one as well because I see that practice very useful. Yet I guess it's somewhat a bad practice, not sure. What do you think?

Reply:
This would prevent returning a value from the promise. (not the promise)
Rule explanation says - "The return value of the executor is ignored. Returning a value from an executor function is a possible error because the returned value cannot be used and it doesn't affect the promise in any way"
Which makes sense, so I have no problem with this rule.

Reply:
I guess main point here is this
```
new Promise((resolve, reject) => {
    if (someCondition) {
        return defaultResult;
    }
    getSomething((err, result) => {
        if (err) {
            reject(err);
        } else {
            resolve(result);
        }
    });
});
```
if you do return default value, promise is never handled at all (nor resolves and nor rejected)
same here:
```
new Promise(() => {
    return 1;
});
```
So the rule enforces that promise is always resolved or rejected

Reply:
But I guess javascript garbage collector would be smart enough to collect pending promises once they have no references, right?

Reply:
Even though the garbage collector handles it, it would be a nice to have standard for codebases.

Reply:
So it's not about returning a promise in a `.then()`. Then sure, why not.

Reply:
A somewhat similar implementation can be seen here - https://github.com/Countly/countly-server/blob/bbf1aa88676b4024ff89dc37069b0ec987c8034c/api/utils/requestProcessor.js#L2425
The promise is called recursively, and once the most recent promise is resolved, the others behind it are neither resolved nor rejected.

Reply:
Oh, good point BTW. It'd prevent returning from the executor function scope, right? 

```
new Promise((res, rej) => {
    if (something.notExists()) {
        return res(); // <<<<< error
    }

    res('something exists');
});
````

Reply:
yes

Reply:
Then I'm back against this :)

Reply:
It does not prevent return from executor scope in this case (as flow control), just make sure all routes resolve or reject
```
new Promise((resolve, reject) => {
    if (someCondition) {
        resolve(defaultResult);
        return;
    }
    getSomething((err, result) => {
        if (err) {
            reject(err);
        } else {
            resolve(result);
        }
    });
});
```

Reply:
As per my understanding, this rule won't check if we are resolving a promise or not.
But rather check for the returns from the promise.
The rule title says - Disallow returning values from Promise executor functions
So it will throw an error if a value is returned from within the promise.
You can still return but without a value bcz a returned value from promise is meaningless.
.then just gets the resolved or rejected values afterall.

Reply:
true, so it won't guard against cases of unhandled promises
I guess then we can turn it off

Reply:
reiterating on this actually :D
so does returning value from executor even make sense? it would not be used anywhere, right, only resolved value is passed returned outside of promise. So it actually prevents you from accidentally thinking you are returning value but you don't

And for @iartem you simply need to do this:
```
new Promise((res, rej) => {
    if (something.notExists()) {
        res();
        return;
    }

    res('something exists');
});
```